### ISSUE TITLE
Enabling Chroot jail domain Unix user, breaks SFTP inaccessible in Centos 8

### ISSUE THREAD

**Issue by dpark77** [virtualmin/virtualmin-gpl#352]
Administration options -> Edit Owner limits ,
Enabling Chroot jail domain Unix user causes SFTP to break and become inaccessible in FIlezilla.

Error:	Could not connect to server (latest version of Filezilla)

Error: Received unexpected end-of-file from SFTP server (older version of Filezilla)
Error: Could not connect to server

I didn't test it with other SFTP apps yet but googling the logged error messages, I found an identical issue at link below and that this may be a bug in systemd?

[https://www.mail-archive.com/debian-bugs-dist@lists.debian.org/msg1668252.html]

BTW, this happens in Centos 8, 
I've confirmed it works just fine in Centos 7.

Error log:
pam_systemd(sshd:session): Failed to create session: Failed to add required mount "/home/chroot/164164858199/./home/name

**Comment by jcameron** [comment]
Does `/home/chroot/164164858199` exist?

**Comment by dpark77** [comment]
Yes, I tried everything chown root/username permission,
even tried completely deleting /home/chroot/~ folder and re-enabling chroot jail.

It only works in filezilla when chroot jail isn't selected.

**Comment by jcameron** [comment]
What about other protocols, like just regular FTP? Or SCP?

**Comment by dpark77** [comment]
Just tested on FTP, same result as SFTP.
I don't use SCP so not sure about that.

**Comment by jcameron** [comment]
Which FTP server are you running? Is it the default ProFTPd that Virtualmin installs?

**Comment by dpark77** [comment]
Yes, I tested it on ProFTPd that comes with virtualmin in Centos 8.

**Comment by jcameron** [comment]
I just tested this with a CentOS system and a domain configured to be jailed, but I wasn't able to reproduce this problem..

**Comment by dpark77** [comment]
That's strange..

I'm on CentOS Linux 8.5.2111 system with the latest virtualmin version.

The following error message is all I see in /var/log/secure

Jan 12 21:21:23  sshd[661503]: Accepted password for dname from xxx.xxx.xx.xx port 24990 ssh2
Jan 12 21:21:23  sshd[661503]: pam_systemd(sshd:session): Failed to create session: Failed to add required mount "/home/chroot/1641648581909/./home/dname": Success
Jan 12 21:21:23  sshd[661503]: pam_unix(sshd:session): session opened for user dname by (uid=0)
Jan 12 21:21:23  sshd[661503]: pam_unix(sshd:session): session closed for user dname

It's having trouble with mounting the above folder path? 

That's why I thought this has something to do with dot(.) between the slashes( / . / ) as mentioned here: 
https://www.mail-archive.com/debian-bugs-dist@lists.debian.org/msg1668252.html

Do you have any idea? whether this could be the reason why it's failing to authenticate or it's something else.

**Comment by jcameron** [comment]
The dot is necessary - some tools use it to find out which part of the path is the chroot directory.

**Comment by dpark77** [comment]
I tried it on many other FTP apps and all seem to be having this same problem.

systemd-logind[1007]: Failed to start session scope session-304.scope: Failed to add required mount "/home/chroot/1641648581329/./home/dname": Success

It works well in other OS like Debian and Centos 7.

**Comment by jcameron** [comment]
Oh .... what line did you change exactly?

**Comment by iliarostovtsev** [comment]
This issue is the result of missing exec bit for a group for the home directory.

For instance, if your domain **`domain.com`** has a home at **`/home/domain.com`**, make sure that the permissions for the home dir of this domain are set at least to 750, i.e.:

```
chmod 0750 /home/domain.com
```

At least, that was my issue when jail was enabled -- I couldn't FTP login with default settings (after creating a new user using **Edit Users: Mail and FTP Users** page), as the permissions were set as follows (for **`centos8-gpl-2`** domain):

```
[root@debug-centos8-gpl home]# ls -lsa
total 0
0 drwxr-xr-x.  5 root          root           60 Jan 17 20:45 .
0 dr-xr-xr-x. 17 root          root          244 Jan 14 21:28 ..
0 drwxr-x---. 11 centos8-gpl   centos8-gpl   272 Jan 14 21:32 centos8-gpl
0 drwx------  10 centos8-gpl-2 centos8-gpl-2 261 Jan 17 20:43 centos8-gpl-2
0 drwxr-xr-x   3 root          root           28 Jan 17 20:45 chroot

```

**Comment by iliarostovtsev** [comment]
(although this only true for FTP connection using ProFTPd service).

**Comment by iliarostovtsev** [comment]
Interestingly that SFTP works just fine on Debian 10, while fails on CentOS 8 when jail is enabled.

Although, FTP user login fails on all distros with or without enabled jail unless adding group exec bit for a home dir (as described above).

Jamie, should we fix that?

**Comment by dpark77** [comment]
In sshd_config,

I just replaced

Subsystem sftp /usr/lib/openssh/sftp-server

with

Subsystem sftp internal-sftp


I don't know what would be the consequence in terms of security or other issues by modifying this part
but at least it's working for me now :)

Thanks for your time.

**Comment by jcameron** [comment]
@iliarostovtsev Virtualmin does set the group exec bit on `/home/username` for all new domains by default though.

**Comment by jcameron** [comment]
Actually, what does the `homedir_perms` line in `/etc/webmin/useradmin/config` contain on your system?

**Comment by iliarostovtsev** [comment]
Yeah, I have checked that before. All set correctly:

```
root@debian10-pro:/home# cat /etc/webmin/useradmin/config | grep homedir_perms
homedir_perms=0750
```

Although, when you start creating a new virtual server, the permissions are set correctly at first (note the new domain here is **`debian10-pro-2`**):

```
root@debian10-pro:/home# ls -lsa
total 24
4 drwxr-xr-x  6 root           root           4096 Jan 18 09:56 .
4 drwxr-xr-x 19 root           root           4096 Dec 16 12:56 ..
4 drwxr-x---  2 debian         users          4096 Aug 29 16:27 debian
4 drwxr-x--- 16 debian10-pro-1 debian10-pro-1 4096 Dec  6 11:13 debian10-pro-1
4 drwxr-x---  9 debian10-pro-2 debian10-pro-2 4096 Jan 18 09:56 debian10-pro-2
4 drwxr-xr-x  2 user           user           4096 Jul 13  2021 user
```

... however, in the middle of domain creation, it get's reset by something and when domain setup is finished it looks like this:

```
root@debian10-pro:/home# ls -lsa
total 24
4 drwxr-xr-x  6 root           root           4096 Jan 18 09:56 .
4 drwxr-xr-x 19 root           root           4096 Dec 16 12:56 ..
4 drwxr-x---  2 debian         users          4096 Aug 29 16:27 debian
4 drwxr-x--- 16 debian10-pro-1 debian10-pro-1 4096 Dec  6 11:13 debian10-pro-1
4 drwx------ 13 debian10-pro-2 debian10-pro-2 4096 Jan 18 09:56 debian10-pro-2
4 drwxr-xr-x  2 user           user           4096 Jul 13  2021 user

```

---

I have taken a deeper look and the culprit of the problem is "Setup Apache SSL website" feature, which calls **`create_ssl_certificate_directories`** with running set permissions on existing dir, changing it to 700.

I have made two PRs - choose whichever seems more correct to you! :smile: 

PR #1 - https://github.com/virtualmin/virtualmin-gpl/pull/356

PR #2 - https://github.com/virtualmin/virtualmin-gpl/pull/357

**Comment by jcameron** [comment]
They both seem correct!

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.
